lib/util/glob-util.js

```diff
@@ -0,0 +1,144 @@
+/**
+ * @fileoverview Utilities for working with globs and the filesystem.
+ * @author Ian VanSchooten
+ * @copyright 2015 Ian VanSchooten. All rights reserved.
+ * See LICENSE in root directory for full license.
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+var debug = require("debug"),
+    fs = require("fs"),
+    glob = require("glob"),
+    shell = require("shelljs"),
+
+    IgnoredPaths = require("../ignored-paths");
+
+debug = debug("eslint:glob-util");
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Checks if a provided path is a directory and returns a glob string matching
+ * all files under that directory if so, the path itself otherwise.
+ *
+ * Reason for this is that `glob` needs `/**` to collect all the files under a
+ * directory where as our previous implementation without `glob` simply walked
+ * a directory that is passed. So this is to maintain backwards compatibility.
+ *
+ * Also makes sure all path separators are POSIX style for `glob` compatibility.
+ *
+ * @param {string[]} [extensions] An array of accepted extensions
+ * @returns {Function} A function that takes a pathname and returns a glob that
+ *                     matches all files with the provided extensions if
+ *                     pathname is a directory.
+ */
+function processPath(extensions) {
+    var suffix = "/**";
+
+    if (extensions) {
+        if (extensions.length === 1) {
+            suffix += "/*." + extensions[0];
+        } else {
+            suffix += "/*.{" + extensions.join(",") + "}";
+        }
+    }
+
+    /**
+     * A function that converts a directory name to a glob pattern
+     *
+     * @param {string} pathname The directory path to be modified
+     * @returns {string} The glob path or the file path itself
+     * @private
+     */
+    return function(pathname) {
+        var newPath = pathname;
+
+        if (shell.test("-d", pathname)) {
+            newPath = pathname.replace(/[\/\\]$/, "") + suffix;
+        }
+
+        return newPath.replace(/\\/g, "/").replace(/^\.\//, "");
+    };
+}
+
+//------------------------------------------------------------------------------
+// Public Interface
+//------------------------------------------------------------------------------
+
```

Comment:
Might be a good idea to create a rule for this. Although this would be hard to do.